types/state/stateTxn.go

```diff
@@ -0,0 +1,334 @@
+
+	//return the state of the BPT for the state of the block
+	rh := types.Bytes(tx.state.RootHash()).AsBytes32()
+	tx.state.logInfo("WriteStates", "height", blockHeight, "hash", logging.AsHex(rh))
+	return tx.state.RootHash(), currentStateCount, nil
+	panic("TODO")
+}
+
+func (tx *DBTransaction) writeChainState(group *sync.WaitGroup, mutex *sync.Mutex, mm *managed.MerkleManager, chainId types.Bytes32) {
+	defer group.Done()
+
+	// We get ChainState objects here, instead. And THAT will hold
+	//       the MerkleStateManager for the chain.
+	//mutex.Lock()
+	currentState := tx.state.updates[chainId]
+	//mutex.Unlock()
+
+	if currentState == nil {
+		panic(fmt.Sprintf("Chain state is nil meaning no updates were stored on chain %X for the block. Should not get here!", chainId[:]))
+	}
+
+	//add all the transaction states that occurred during this block for this chain (in order of appearance)
+	for _, txn := range currentState.txId {
+		//store the txHash for the chains, they will be mapped back to the above recorded tx's
+		tx.state.logInfo("AddHash", "hash", logging.AsHex(txn))
+		mm.AddHash(managed.Hash(*txn))
+	}
+
+	if currentState.stateData != nil {
+		//store the MD root for the state
+		mdRoot := mm.MS.GetMDRoot()
+		if mdRoot == nil {
+			//shouldn't get here, but will reject if I do
+			panic(fmt.Sprintf("shouldn't get here on writeState() on chain id %X obtaining merkle state", chainId))
+		}
+
+		//store the state of the main chain in the state object
+		currentState.stateData.MDRoot = types.Bytes32(*mdRoot)
+
+		//now store the state object
+		chainStateObject, err := currentState.stateData.MarshalBinary()
+		if err != nil {
+			panic("failed to marshal binary for state data")
+		}
+
+		mutex.Lock()
+		tx.state.GetDB().Key(bucketEntry.AsString(), chainId.Bytes()).PutBatch(chainStateObject)
+		// The bpt stores the hash of the ChainState object hash.
+		tx.state.bpt.Bpt.Insert(chainId, sha256.Sum256(chainStateObject))
+		mutex.Unlock()
+	}
+	//TODO: figure out how to do this with new way state is derived
+	//if len(currentState.pendingTx) != 0 {
+	//	mdRoot := v.PendingChain.MS.GetMDRoot()
+	//	if mdRoot == nil {
+	//		//shouldn't get here, but will reject if I do
+	//		panic(fmt.Sprintf("shouldn't get here on writeState() on chain id %X obtaining merkle state", chainId))
+	//	}
+	//	//todo:  Determine how we purge pending tx's after 2 weeks.
+	//	s.bpt.Bpt.Insert(chainId, *mdRoot)
+	//}
+}
+
+func (tx *DBTransaction) logInfo(msg string, keyVals ...interface{}) {
+	if tx.state.logger != nil {
+		// TODO Maybe this should be Debug?
+		tx.state.logger.Info(msg, keyVals...)
+	}
+}
+
+func (tx *DBTransaction) GetDB() *database.Manager {
+	return tx.state.db
+}
+
+func (tx *DBTransaction) Sync() {
+	tx.state.sync.Wait()
+}
+
+func (tx *DBTransaction) RootHash() []byte {
+	h := tx.state.bpt.Bpt.Root.Hash // Make a copy
+	return h[:]                     // Return a reference to the copy
+}
+
+func (tx *DBTransaction) writeBatches() {
```

Comment:
Instead of adding a copy of `writeBatches`, change the receiver of the original.